README.md
Archey :: OS X
--------------

Simple bash script that displays the Apple logo and some basic information.

---

><span style="font-size: small; color: #555;">_Credits: This is a modified version of: "[archey-osx][jfa]" by [Josh Finnie][jfg], who made the original bash script. The name Archey is of course based on [Archey][dja], by [djmelek][djm]._</span>


### Install

Copy the files to `/usr/local/bin`, and give them execute permission.

```
sudo cp archey{,.s} /usr/local/bin
sudo chmod +x /usr/local/bin/archey{,.s}
```

### Usage

To load at start. Add in your [.bashrc][inst] _(or .bash_profile)_. Example:

```
# Load archey (if installed)
[[ `which archey` && $UID != 0 ]] && archey
```
In top of the file there are 2 settings...

```
diskWarning=1;             # bool
diskWarningThreshold=20;   # % left on disk left
```
_The disk warning is just a visual blinking warning (of <span style="color: #900;"> > 80%</span>)._ 



#### archey.s

`archey.s` is a screenshot-plugin I made xtra to use with `archey`, but can be used as a stand alone script as well.

To use it with `archey`:

```
archey -s		# screenshot ... countdown from 5 sec
archey -sw		# w = window mode.
```
As a stand alone script:

```
archey.s		# screenshot ... countdown from 5 sec
archey -w		# w = window mode.
```

In top of the file there are a few settings:

```
# Settings
# ------------------------------------------------------------
forceDir=0;         # Default: 0. Will force to use "our"
                    # directory rather than the normal one.

asDir="$HOME/ArcheyScreenshots";
asType='png';
asPrefix='ArcheyOSX';
asDate=`date "+_%Y-%m-%d_at_%H.%M.%S"`;
```

There's also an `-f` to force the use of `ArcheyScreenshots`.

```
archey.s -w -f    // example
```



### Screenshot

![][sc]

<span style="font-size: small; color: #555;">_* The picture has a slighlty pale tone since I'm using a semi transparent background in my Terminal (my theme/prefs (“myTerm”) can be found in [dotfiles][mt])_</span>



<!-- Markdown: Links & Images -->
[inst]: https://github.com/iEFdev/dotfiles/blob/master/.bashrc#L87-L88

[mt]: https://github.com/iEFdev/dotfiles
[jfa]: https://github.com/joshfinnie/archey-osx
[jfg]: http://joshfinnie.github.io/

[dja]: https://github.com/djmelik/archey
[djm]: https://github.com/djmelik

[sc]: https://raw.githubusercontent.com/iEFdev/Archey-OS-X/master/screenshot.png "Screenshot of Archey"

CHANGES.md
# Changes

Changes that have been made between the versions.

- [ ] _todo: move this file to Wiki: Changelog._

- - -

> **Latest version:** v0.7.0-beta

### Changes since: `v0.6.1` (Nov 21, 2016)

-	Fixed better compare of floats in version nr (`free`).
-	Replaced `optparse` (deprecated) with `argparse`.
-   Code has been rn with pylint. Only 1 report left, and that's the filename - which I intend to keep.
-   Optimized/refactored some code
-   Removed `platform.linux_distribution()`. It's deprecated
    and flagged for removal, so...
-   A new function to detect distro name and full distro name. No using `/etc/{{os,lsb}-release,issue}` and `lsb_release [-si|-sd]`. In that way everything can take place in on function.
-   A new function to parse the release files
-   New (xtra) dict - separating distros and colors, to 1 color for multiple dists
-   RAM and Disk output is formatted a bit different
-   Changed the %-levels for disk-colors
-   Added `simfs` and `tmpfs` as mentioned here:
    - <https://github.com/djmelik/archey/pull/38>
    - <https://github.com/djmelik/archey/issues/39>


> **Note:** v0.6.1 = last version of Archey-OS-X. _Next version is “Archey X” (X compatible)_


### Changes since: v0.6.0
-	Misc cleaning, detabbing, and small fixes
-	Added 10.12 Sierra.
-   _Note: v0.6.1 was the last version that is OS X-only._


### Changes since: v0.5.0

-	New additions in wmDict. “Quartz” will be displayed as default, and any other will be picked up if installed.
-	Macports and Fink was added to Packages, ~~and 'pkgutil' for OS X programs (installer)~~. The combined number of all will be displayed.
-	New feat: Fancy Names. Displaying version names instead of build number _(with a fallback to buildnumber if not in dict)_.
-	Rewrote the Uptime class, using the built in timedelta to format the code.
-	Reduced the usage of Popen, and using the built in libraries instead (username, hostname etc).
-	Better output of RAM. Used/free sumup exactly as Activiy Monitor now (was a small diff before).



### Changes since: v0.4.0

-	Script is detecting “Darwin” with `uname -s`, not by checking if a file exists.
-	`logosDict` moved to the end, and to pick up value from variable (`distro`)
-	New class (`Utils`), to clean up the code.
-	Made to work with `python` and `python3`. Original script is using `python3` but standard version in OS X is `python` _(2.7)_ for those who doesn't have a newer one.
-	Removed som redundant code.
-	Added a simple option with  using `optparse`: `--version`
	_(incl the standard: `-h` and `--help`)_..
-	`brand` defaults to “None”. Self detecting. Only needed to use when changing color scheme.
-	Fixed an issue with decimals in RAM, which showed up when testing different versions of python. _(Used `math.trunc()`)_
-	With `archey.bash` gone, there's an xtra logo with the Apple to use. Default is the “Arch” logo.
-	`archey.{bash,s}` is removed and branched off to `bash-version`.


archey.s
#!/usr/bin/env bash
# ------------------------------------------------------------
#
# /usr/local/bin/archey
#
#
# Plugin for Archey :: OS X
#
# Description:  Built to add a screenshot feature to Archey-OS X,
#               but can run as a stand alone script.
#
# Usage:
# - with archey (bash)  archey -s       archey -sw
# - as stand alone      archey.s        archey.s -w
#
# Version:      1.0.0
# Author:       Eric F (iEFdev)
# Copyright:    (c) 2015, Eric F
# Licence:      GNU General Public License v3
#


# Settings
# ------------------------------------------------------------
forceDir=0;         # Default: 0. Set to 1 will force to use
                    # "our" directory rather than the normal one.

asDir="$HOME/ArcheyScreenshots";
asType='png';
asPrefix='Archey-OSX';
asDate=`date "+_%F_at_%X" | sed 's/\:/\./g'`;


# Directory
# ------------------------------------------------------------
# Where to store the image

# Force it with a handler ("f"), when the setting is set to 0.
if [[ $forceDir == 0 ]] && [[ $@ =~ [f] ]]; then forceDir=1; fi

# First, get the user path to screenshots...
# If there's no default location. Create a folder and use it.
uDir=`defaults read com.apple.screencapture location | sed 's/[ \/]*$//'`;
if [[ $uDir != '' ]] && [[ $forceDir == 0 ]]; then
	asDir="$uDir";
elif [[ $uDir == '' ]] || [[ $forceDir == 1 ]]; then
	if [[ ! -d $asDir ]]; then
		mkdir -p $asDir;
	fi
fi


# Function _help()
# Perhaps someone find it useful :)
function _help() {
	if [[ $thisArchey == 'archey.s' ]]; then
			uUse='[-w]';
			hS='default';
			hSW='-w ';
	else
			uUse='[-sw]';
			hS='-s';
			hSW='-sw';
	fi

	cat <<EOF

Usage: $thisArchey $uUse[-f][--help]

	$hS	  Takes a screenshot of the full display

	$hSW	  Triggers a window mode.
	    	  (Unfortunately not automatically. Needs a click.)

	-f	  Will force to use the folder defined as: \"\$asDir=[DIR]\"
	    	  in the settings, if \"forceDir=0\". If the folder doesn't
	    	  exists, it will try to create it.

	- - - - -

	When the picture is taken it will be revealed in Finder.

EOF
}


# Function _counter
# 5.. 4.. 3.. 2.. 1..
function _counter() {
	echo -n "Taking a screenshot in: "
	for i in {5..1}.. '[Click!]'; do
		echo -n "$i ";
		sleep 1;
	done;
	echo;
}


# More variables
asImg="$asDir/$asPrefix$asDate.$asType";
asMode="$1";
asSec=1;
thisArchey=`basename "$0"`;


# Ok, who's who..?
# ------------------------------------------------------------

# archey.s
if [[ $asMode == '' ]] && [[ $thisArchey == 'archey.s' ]]; then
	_counter;
	asOpts='xo';
elif  [[ $asMode != '' ]]; then
	case $asMode in

		# archey, archey.s
		--help)
			_help;
			;;

		# archey
		-s)
			if [[ $thisArchey == 'archey.s' ]]; then $0 *;
			else
				_counter;
				asOpts='xo';
			fi
			;;

		# archey, archey.s
		-w|-sw)
			if [[ $thisArchey == 'archey.s' && $asMode != '-w' ]]; then $0 *;
			else
				asOpts='xow';
				asSec=0;
			fi
			;;

		# Oops
		*)
			if [[ $thisArchey == 'archey.s' ]]; then uUse='[-w][-f]'; else uUse='[-sw][-f]'; fi
			echo -e " :» $thisArchey:\tillegal option. ($@)";
			echo -e " :» usage:\t$thisArchey $uUse[--help]";
			exit;
			;;

	esac
fi

# ...and here we go.
if [[ $asOpts ]]; then
	asOpt="-T $asSec -t $asType -$asOpts";
	screencapture $asOpt $asImg && open -R $asImg;
fi

# Bye bye...
exit;


archey.bash
#!/usr/bin/env bash
# ------------------------------------------------------------
#
# /usr/local/bin/archey
#
#
# Archey :: OS X
#
# Description:  Tool (bash script) to display system information.
#               Includes an option/plugin to take screenshots.
#
# Version:      1.0.0
# Author:       Eric F (iEFdev)
# Copyright:    (c) 2015 Eric F
# Licence:      GNU General Public License v3
#
# Author notes: This is a modified version of: "archey-osx"
#               by Josh Finnie.
#                   ↳ https://github.com/joshfinnie/archey-osx
#                   ↳ http://joshfinnie.github.io/
#


# Settings
# ------------------------------------------------------------
# diskWarning            (default: 1) with green/red colors
# diskWarningThreshold   (default: 20) % of diskspace left
diskWarning=1;
diskWarningThreshold=20;


# Colors
# ------------------------------------------------------------
# 2 first lines = the apple and the labels

b_c="\033[1;36m";		# b_c = ascii:bold color
n_c="\033[0;36m";		# n_c = ascii:normal color
g_c="\033[0;32m";		# g_c = green color
r_c="\033[0;31m";		# r_c = red color
w_c="\033[5;31m";		# w_c = red color (warning) (5; = blinking)
e_c="\033[0m";			# e_c = empty color (eg reset)


# Archey variables
# ------------------------------------------------------------

# User
user=$(whoami);

# Hostname
hostname=$(uname -n);

# Distro
distro="$(sw_vers -productName) $(sw_vers -productVersion) ($(sw_vers -buildVersion))";

# Kernel
kernel=$(uname -srm);

# Uptime: Example: "35 min" or "1 day, 12:35" or "1 day, 10 min"
uptime=$(uptime | sed 's/  / /g' | sed 's/.*up \([^,].*\),.*/\1/' | sed 's/\([^,].*\),.*/\1/');

# Shell and Terminal
shell="$SHELL";
terminal="$TERM";

# Packages
# Display nice if Homebrew isn't installed
if [[ `which brew` ]] && return 1; then
	packages="`brew list -l | wc -l | awk '{ print $1 }'`";
else
	packages="n/a";
fi

# CPU (info)
cpu=$(sysctl -n machdep.cpu.brand_string | sed -e 's/  */ /g' | sed 's/CPU\ //');


# Memory section
# ------------------------------------------------------------
# Display: (used = red) (free = green) total
#
memSize=$(sysctl -n hw.memsize);
memTotalRam="$((memSize/1000000000)) GB";

# Using vm_stat to get all the memory
memFree=$(vm_stat | grep "free" | awk '{ print $3 }' | sed 's/\.//');
memInactive=$(vm_stat | grep "Pages inactive" | awk '{ print $3 }' | sed 's/\.//');
memSpec=$(vm_stat | grep "speculative" | awk '{ print $3 }' | sed 's/\.//');
memWired=$(vm_stat | grep "wired" | awk '{ print $4 }' | sed 's/\.//');
memActive=$(vm_stat | grep "Pages active" | awk '{ print $3 }' | sed 's/\.//');

# Used memory
# These 3 are matching "1234M used" in: top (and Activity Monitor)
memUsed=$((($memWired+$memActive+$memInactive)*4096/1024/1024));

# Total free
memTotalFree=$((($memFree+$memSpec)*4096/1024/1024));

# [note]
# I's a little bit of a mess, and slightly confusing. Size of total memory ($memTotalRam) is easy to
# get, but adding up what's used seems it's using MiB instead of MB. Haven't figured out
# a way to solve it. Did try a few different settings, but any ideas/suggestions/improvements
# are welcome.

# RAM:	(used) (free) total
ram=`echo -e "$r_c($memUsed MB)$e_c $g_c($memTotalFree MB)$e_c / $memTotalRam"`;


# Function _disk()
# ------------------------------------------------------------
# Display: used/N% of total (N available) (disk)
#
function _disk() {

	if [[ $diskWarning == 0 ]]; then

		# Without colors...
		disk=`df -Hl | head -2 | tail -1 | awk '{print $3"/"$5 " of " $2 " (" $4 " available) (" $1 ")"}'  | sed 's/\/dev\///g'`;
	else
		# DiskWarning will show some red text when the space
		# left on disk is below the setting.

		# used size
		uSize=`df -Hl | head -2 | tail -1 | awk '{ print $3}'`;

		# used size (%)
		pSize=`df -Hl | head -2 | tail -1 | awk '{ print $5}'`;

		# total size
		totSize=`df -Hl | head -2 | tail -1 | awk '{ printf "of %-1s", $2}'`;

		# available size
		aSize=`df -Hl | head -2 | tail -1 | awk '{ printf "(%-1s available)", $4}'`;

		# diskNsN
		mPoint=`df -Hl | head -2 | tail -1 | awk '{printf "(%-1s)", $1}'  | sed 's/\/dev\///g'`;

		# Ok... If the disk is more than 80% full?
		if [[ $pSize > $((100-$diskWarningThreshold)) ]]; then
			pSize="$w_c$pSize$e_c";		# Blink % + red color
			aSize="$r_c$aSize$e_c";		# red color
		else
			pSize="$g_c$pSize$e_c";		# % in green color
		fi

		# Sum up ...
		disk="$uSize/$pSize $totSize $aSize $mPoint";
	fi
}
# load
_disk;


# Archey :: OS X
# ------------------------------------------------------------
echo -e "
$b_c                        ##  $n_c         User:$e_c      $user
$b_c                     #####  $n_c         Hostname:$e_c  $hostname
$b_c                    #####   $n_c         Distro:$e_c    $distro
$b_c                  #####     $n_c         Kernel:$e_c    $kernel
$b_c                  ###       $n_c         Uptime:$e_c    $uptime
$b_c           #####      ##### $n_c         Shell:$e_c     $shell
$b_c       #####################$n_c#        Terminal:$e_c  $terminal
$b_c     ###################$n_c#######      Packages:$e_c  $packages
$b_c    ################$n_c#########        CPU:$e_c       $cpu
$b_c    #############$n_c###########         RAM:$e_c       $ram
$b_c    ##########$n_c#############          Disk:$e_c      $disk
$b_c    #######$n_c#################   $e_c
$b_c    #####$n_c##################### $e_c
$b_c     ###$n_c#######################$e_c
$b_c      #$n_c####################### $e_c
$n_c        #####################  $e_c
$n_c         ##################    $e_c
$n_c           ####      ####      $e_c
";


# Plugins
# ------------------------------------------------------------

# Import the archey.s plugin (if installed)
# The plugin will add a screenshot feature
if [ -f `which archey.s` ] && [[ $1 != '' ]]; then
	. `which archey.s`;
fi

exit;


archey.py
#!/usr/bin/env python
# ------------------------------------------------------------
#
# /usr/local/bin/archey
#
#
# Archey :: OS X
#
# Description:  Archey is a simple system information tool
#               written in Python.
#
# Version:      0.4.0
# Author:       Eric F (iEFdev)
# Copyright:    (c) 2015 Eric F
# Licence:      GNU General Public License v3
#
#
# Author notes:
# ------------------------------------------------------------
#
# This is a forked (ported) version of the original Archey by Melik Manukyan,
# rewritten for OS X.
#
# Original author(s) and copyright:
#
#	Copyright 2010 Melik Manukyan <melik@archlinux.us>
#	Copyright 2010 David Vazgenovich Shakaryan <dvshakaryan@gmail.com>
#
#	ASCII art by Brett Bohnenkamper <kittykatt@silverirc.com>
#	Changes Jerome Launay <jerome@projet-libre.org>
#	Fedora support by YeOK <yeok@henpen.org>
#
# Distributed under the terms of the GNU General Public License v3.
# See LICENSE or http://www.gnu.org/licenses/gpl.txt for the full license text.
#


# Settings
# ------------------------------------------------------------
# diskWarning            default: 1
# diskWarningThreshold   default: 20 (% of diskspace left)
diskWarning = 1;
diskWarningThreshold = 20;


# OS / Dist (for colors)
# Match a name in in colorDict
brand = 'OS X'


# Import libraries
import os, sys, subprocess, optparse, re, linecache
from subprocess import Popen, PIPE
from optparse import OptionParser
from getpass import getuser
from time import ctime, sleep


# Output
# ------------------------------------------------------------
output = [
	'User',
	'Hostname',
	'Distro',
	'Kernel',
	'Uptime',
	'WindowManager',
	'DesktopEnvironment',
	'Shell',
	'Terminal',
	'Packages',
	'CPU',
	'RAM',
	'Disk'
]


# Dictionaries
# ------------------------------------------------------------
colorDict = {
	'OS X': ['\x1b[0;36m', '\x1b[1;36m'],
	'Arch Linux': ['\x1b[0;34m', '\x1b[1;34m'],
	'Ubuntu': ['\x1b[0;31m', '\x1b[1;31m', '\x1b[0;33m'],
	'Debian': ['\x1b[0;31m', '\x1b[1;31m'],
	'Mint': ['\x1b[0;32m', '\x1b[1;37m'],
	'Crunchbang': ['\x1b[0;37m', '\x1b[1;37m'],
	'Fedora': ['\x1b[0;34m', '\x1b[1;37m'],
	# Sensors: [<red>, <yellow>, <green>, <blinking red>]
	'Sensors': ['\x1b[0;31m', '\x1b[0;32m', '\x1b[0;33m', '\x1b[5;31m'],
	'Clear': ['\x1b[0m']
}

deDict = {
	'Finder': 'Finder'
}

# @todo: check up it there are more window tiling programs for OS X.
wmDict = {
	'ShiftIt': 'ShiftIt',
	'Slate': 'Slate',
	'Spectacle': 'Spectacle',
	'Xmonad': 'Xmonad'
}

logosDict = {brand: '''{color[1]}
{color[1]}               +                {results[0]}
{color[1]}               #                {results[1]}
{color[1]}              ###               {results[2]}
{color[1]}             #####              {results[3]}
{color[1]}             ######             {results[4]}
{color[1]}            ; #####;            {results[5]}
{color[1]}           +##.#####            {results[6]}
{color[1]}          +##########           {results[7]}
{color[1]}         ######{color[0]}#####{color[1]}##;         {results[8]}
{color[1]}        ###{color[0]}############{color[1]}+        {results[9]}
{color[1]}       #{color[0]}######   #######        {results[10]}
{color[0]}     .######;     ;###;`\".      {results[11]}
{color[0]}    .#######;     ;#####.       {results[12]}
{color[0]}    #########.   .########`     {results[13]}
{color[0]}   ######'           '######    {results[14]}
{color[0]}  ;####                 ####;   {results[15]}
{color[0]}  ##'                     '##   {results[16]}
{color[0]} #'                         `#  {results[17]}
\x1b[0m'''
}

# Get processes/names
processes = str(subprocess.check_output(('ps', '-u', getuser(), '-o', 'comm')).split('/')).replace('\\n', '').rstrip('\n')


# Classes
# ------------------------------------------------------------
class Output:
	results = []
	results.extend(['']*(18-len(output)))

	def __init__(self):
		self.distro = self.__detectDistro()

	def __detectDistro(self):
		# OS X file
		if os.path.exists('/System/Library/LaunchDaemons/bootps.plist'):
			return brand
		else:
			sys.exit(1)

	def append(self, display):
		self.results.append('%s%s: %s%s' % (colorDict[self.distro][1], display.key, colorDict['Clear'][0], display.value))

	def output(self):
		print(logosDict[self.distro].format(color = colorDict[self.distro], results = self.results))


class User:
	def __init__(self):

		self.key = 'User'
		self.value = os.getenv('USER')


class Hostname:
	def __init__(self):
		hostname = Popen(['uname', '-n'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')

		self.key = 'Hostname'
		self.value = hostname


class Distro:
	def __init__(self):
		if os.path.exists('/System/Library/LaunchDaemons/bootps.plist'):
			productName = Popen(['sw_vers', '-productName'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
			productVersion = Popen(['sw_vers', '-productVersion'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
			buildVersion = Popen(['sw_vers', '-buildVersion'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')

			distro = '%s %s (%s)' % (productName, productVersion, buildVersion)

		self.key = 'Distro'
		self.value = distro


class Kernel:
	def __init__(self):
		kernel = Popen(['uname', '-srm'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')

		self.key = 'Kernel'
		self.value = kernel


class Uptime:
	def __init__(self):
		getUptime = Popen(['uptime'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
		uptime = re.compile(r'(, )(\d )(.*)').sub(r"", getUptime)
		uptime = re.compile(r'(.*)up(\s+)(.*)').sub(r"\3", uptime)
		uptime = re.sub('  +', ' ', uptime)

		self.key = 'Uptime'
		self.value = uptime


class WindowManager:
    def __init__(self):
        wm = ''
        for key in wmDict.keys():
            if key in processes:
                wm = wmDict[key]
                break

        self.key = 'Window Manager'
        self.value = wm


class DesktopEnvironment:
	def __init__(self):
		de = ''
		for key in deDict.keys():
			if key in processes:
				de = deDict[key]
				break

		self.key = 'Desktop Environment'
		self.value = de


class Shell:
    def __init__(self):

        self.key = 'Shell'
        self.value = os.getenv('SHELL')


class Terminal:
    def __init__(self):

        self.key = 'Terminal'
        self.value = os.getenv('TERM')


class Packages:
    def __init__(self):
		checkBrew = Popen(['which', 'brew'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
		p1 = Popen(['which', 'brew', '-l'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
		packages = len(p1.rstrip('\n').split('\n'))

		self.key = 'Packages'

		if checkBrew:
			self.value = packages
		else:
			self.value = 'n/a'


class CPU:
	def __init__(self):
		cpuinfo = Popen(['sysctl', '-n', 'machdep.cpu.brand_string'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
		cpuinfo = re.sub('  +', ' ', cpuinfo.replace('CPU ', '').rstrip('\n'))

		self.key = 'CPU'
		self.value = cpuinfo


# Memory information
# ------------------------------------------------------------
# Display: (used = red) (free = green) total
#
class RAM:
	def __init__(self):
		memSize = Popen(['sysctl', '-n', 'hw.memsize'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
		memTotalRam = int(memSize) / int(1000000000)

		vmStat = Popen(['vm_stat'], stdout=PIPE).communicate()[0].decode('Utf-8').rstrip('\n')
		vmStat = re.sub('  +', ' ', vmStat.replace('.', '').rstrip('\n'))

		# Using vm_stat to get all the memory
		memFree = vmStat.splitlines()[1].split()[2]
		memActive = vmStat.splitlines()[2].split()[2]
		memInactive = vmStat.splitlines()[3].split()[2]
		memSpeculative = vmStat.splitlines()[4].split()[2]
		memWired = vmStat.splitlines()[5].split()[3]

		# Used memory
		# These 3 are matching "1234M used" in: top (and Activity Monitor)
		memUsed = (int(memWired) + int(memActive) + int(memInactive)) * int(4096) / int(1024) / int(1024)

		# Total free
		memTotalFree = (int(memFree) + int(memSpeculative)) * int(4096) / int(1024) / int(1024)

		# [note]
		# I's a little bit of a mess, and slightly confusing.
		# Size of total memory (memTotalRam) is easy to get, but adding up what's
		# used seems it's using MiB instead of MB. Haven't figured out a way to solve it.
		# Did try a few different settings, but any ideas/suggestions/improvements are welcome.

		# RAM:	(red=used) (green=free) / total
		ramdisplay = '%s(%s MB) %s(%s MB)%s / %s GB' % (colorDict['Sensors'][0], memUsed, colorDict['Sensors'][1], memTotalFree, colorDict['Clear'][0], memTotalRam)

		self.key = 'RAM'
		self.value = ramdisplay


# Disk information
# ------------------------------------------------------------
# Display: used/N% of total (N available) (disk)
#
class Disk:
	def __init__(self):
		p1 = Popen(['df', '-lH'], stdout=PIPE).communicate()[0].decode("Utf-8")
		total = p1.splitlines()[1]
		used = total.split()[2]
		size = total.split()[1]
		available = total.split()[3]
		capacity = total.split()[4]
		disknr = total.split()[0].replace('/dev/', '')
		usedpercent = int(capacity.replace('%', ''))
		leftpercent	= (100 - int(usedpercent))

		if usedpercent <= 33:
			diskcolor = 1
		if usedpercent > 33 and usedpercent < 67:
			diskcolor = 2
		if usedpercent >= 67:
			diskcolor = 0
		if diskWarning == 1 and leftpercent <= int(diskWarningThreshold):
			diskcolor = 3

		disk = '%s/%s%s%s of %s (%s available) (%s)' % (used, colorDict['Sensors'][diskcolor], capacity, colorDict['Clear'][0], size, available, disknr)

		self.key = 'Disk'
		self.value = disk


classes = {
	'User': User,
	'Hostname': Hostname,
	'Distro': Distro,
	'Kernel': Kernel,
	'Uptime': Uptime,
	'WindowManager': WindowManager,
	'DesktopEnvironment': DesktopEnvironment,
	'Shell': Shell,
	'Terminal': Terminal,
	'Packages': Packages,
	'CPU': CPU,
	'RAM': RAM,
	'Disk': Disk
}


out = Output()
for x in output:
	out.append(classes[x]())
out.output()
